benchmarks/pipe.py

```diff
@@ -288,42 +169,67 @@ def train(lm_dataloader, model, criterion, optimizer, vocab_size, args):
             total = total.cuda()
         torch.distributed.all_reduce(total, group=model.group)
         logging.info(
-            f"training model, #prams = {num_params}, group: {model.group.rank()}, grank:"
+            f"training model, #params = {num_params}, group: {model.group.rank()}, grank:"
             f" {torch.distributed.get_rank()}, sizes {model.group.size()}"
         )
         torch.distributed.barrier()
         if model.group.rank() == 0:
             logging.info(f"total #prams = {total.item()}")
     else:
-        logging.info(f"training model, #prams = {num_params}")
-    vocab_size = 10000  # FIXME
-    total_loss = 0.0
-    start_time = time.time()
-    word_counter = 0
+        logging.info(f"training model, #params = {num_params}")
 
-    optimizer = optimizer(model)
 
-    def get_first_device(model):
-        if isinstance(model, DDP):
-            model = model.module
+def get_first_device(model):
+    if isinstance(model, DDP):
+        model = model.module
 
-        if not torch.cuda.is_available():
-            return torch.device("cpu")
-        if model.devices:
-            return model.devices[0]
-        else:
-            return torch.cuda.current_device()
+    if not torch.cuda.is_available():
+        return torch.device("cpu")
+    if model.devices:
+        return model.devices[0]
+    else:
+        return torch.cuda.current_device()
 
-    def get_last_device(model):
-        if isinstance(model, DDP):
-            model = model.module
 
-        if not torch.cuda.is_available():
-            return torch.device("cpu")
-        if model.devices:
-            return model.devices[-1]
-        else:
-            return torch.cuda.current_device()
+def get_last_device(model):
```

Comment:
can merge with the last function into get_first_or_last_device to avoid some duplication. Only one line is different between the two anyway.

Reply:
Done.